src/transformStateWithClones.js

```diff
@@ -8,6 +8,40 @@
  */
 function transformStateWithClones(state, actions) {
   // write code here
+  const arrayWithObjects = [];
+  let cloneState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        cloneState = {
+          ...cloneState,
+          ...action.extraData,
+        };
+
+        break;
+
+      case 'removeProperties':
+        cloneState = removeProperties(cloneState, action.keysToRemove);
+
+        break;
+      case 'clear':
+        cloneState = {};
+
+        break;
+    }
+    arrayWithObjects.push({ ...cloneState });
+  }
+
+  return arrayWithObjects;
+}
+
+function removeProperties(newObject, keysToRemove) {
+  for (const key of keysToRemove) {
+    delete newObject[key];
+  }
+
+  return newObject;
 }
```

Comment:
The `removeProperties` function should not modify the object passed to it. Instead, it should create a copy of the object, remove the specified properties from the copy, and then return the copy. This ensures that the original object passed to the function is not mutated, adhering to the checklist item to avoid object mutation.